### PR TITLE
Add DbFunctionParameter support for DbFunctions

### LINKED ISSUE #13752: Add API to configure database types for DbFunction's parameters
See https://github.com/aspnet/EntityFrameworkCore/issues/13736#issuecomment-432871876 for the setup and the conversion.

I attempted to map my own function like this:
```C#
    public static class GeoExtensions
    {
        public static double Distance(this GeoPoint x, GeoPoint y)
            => throw new NotImplementedException();
    }
```
which is based on the GeoPoint mapping in the comment linked above.

My attempt was this:
```C#
modelBuilder.HasDbFunction(
    typeof(GeoExtensions).GetMethod(nameof(GeoExtensions.Distance)),
    b => b.HasTranslation(
        e => new SqlFunctionExpression(
            e.First(),
            "STDistance",
            typeof(double),
            e.Skip(1))));
```
This failed [here](https://github.com/aspnet/EntityFrameworkCore/blob/7e28f43dc79205aae8a74e3e53adf628aa25174f/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L620) because there is no mapping registered for GeoPoint.

This would likely be fixed by #10784, but even then it probably should be possible to make work with a custom mapping.

### LINKED ISSUE #13782: Add model validation to verify DbFunction's parameter types are mapped
See example in #13752. If I try to use this function, which is defined correctly but doesn't work, then all I get is the NotImplementedException. I get no info as to what happened that causes it not to be used.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #13752
Resolves #13782

Replaces #14990

### CODE REVIEW

File: src/EFCore.Relational/Extensions/RelationalModelExtensions.cs [hunk 1]
```diff
@@ -300,21 +301,21 @@ public static IConventionDbFunction RemoveDbFunction([NotNull] this IConventionM
         ///     Returns all <see cref="IDbFunction" />s contained in the model.
         /// </summary>
         /// <param name="model"> The model to get the functions in. </param>
-        public static IReadOnlyList<IDbFunction> GetDbFunctions([NotNull] this IModel model)
-            => DbFunction.GetDbFunctions(model).ToList();
+        public static IEnumerable<IDbFunction> GetDbFunctions([NotNull] this IModel model)
+            => DbFunction.GetDbFunctions((Model)model);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Extensions/RelationalModelExtensions.cs]: We usually call `AsModel()` for `IModel` extensions

Comment (smitpatel) [reply src/EFCore.Relational/Extensions/RelationalModelExtensions.cs]: 😕

File: src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            _parameter = (DbFunctionParameter)parameter;
+        }
+
+        public virtual IMutableDbFunctionParameter Metadata => _parameter;
+
+        /// <inheritdoc />
+        IConventionDbFunctionParameter IConventionDbFunctionParameterBuilder.Metadata => _parameter;
+
+        /// <summary>
+        ///     Specify if this parametere supports Nullability Propagation
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs]: Typo in "parametere"

File: src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+        bool CanSetSupportsNullPropagation(bool supportsNullPropagation, bool fromDataAnnotation = false);
+
+        IConventionDbFunctionParameterBuilder HasStoreType([CanBeNull] string storeType, bool fromDataAnnotation = false);
+
+        bool CanSetStoreType([CanBeNull] string storeType, bool fromDataAnnotation = false);
+
+        IConventionDbFunctionParameterBuilder HasTypeMapping([CanBeNull] RelationalTypeMapping typeMapping, bool fromDataAnnotation = false);
+
+        bool CanSetTypeMapping([CanBeNull] RelationalTypeMapping typeMapping, bool fromDataAnnotation = false);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs]: XML docs? 😉

Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs]: 😢

File: src/EFCore.Relational/Extensions/RelationalModelExtensions.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.Relational/Extensions/RelationalModelExtensions.cs]: @ajcvickers - He suggested me to write this code in first place lol.

Comment (ajcvickers) [reply src/EFCore.Relational/Extensions/RelationalModelExtensions.cs]: Well then I was wrong. AsModel is better.

File: src/EFCore.Relational/Metadata/Conventions/Internal/DbFunctionTypeMappingConvention.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace Microsoft.EntityFrameworkCore.Metadata.Conventions
+{
+    /// <summary>
+    ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+    ///     directly from your code. This API may change or be removed in future releases.
+    /// </summary>
+    public class DbFunctionTypeMappingConvention : IModelFinalizedConvention
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/Internal/DbFunctionTypeMappingConvention.cs]: XML docs

File: src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs]: I can do the xml docs if you want.

File: src/EFCore.Relational/Metadata/IConventionDbFunctionParameter.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace Microsoft.EntityFrameworkCore.Metadata
+{
+    public interface IConventionDbFunctionParameter : IDbFunctionParameter
+    {
+        new IConventionDbFunction Parent { get; }
+
+        IConventionDbFunctionParameterBuilder Builder { get; }
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/IConventionDbFunctionParameter.cs]: Docs

File: src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Builders/IConventionDbFunctionParameterBuilder.cs]: Yay!

**Pull request merged**